arangod/Agency/MoveShard.cpp

```diff
@@ -907,15 +941,83 @@ arangodb::Result MoveShard::abort(std::string const& reason) {
   } else if (res.indices[0] == 0) {
     if (_isLeader) {
       // Tough luck. Things have changed. We'll move on
-      LOG_TOPIC("513e6", INFO, Logger::SUPERVISION) <<
-        "MoveShard can no longer abort through reversion to where it started. Flight forward";
-      finish(_to, _shard, true, "job aborted (4) - new leader already in place: " + reason);
+      LOG_TOPIC("513e6", INFO, Logger::SUPERVISION)
+          << "MoveShard can no longer abort through reversion to where it "
+             "started. Flight forward";
+      moveShardFinish(true, true,
+                      "job aborted (4) - new leader already in place: " + reason);
       return result;
     }
     result = Result(
-    TRI_ERROR_SUPERVISION_GENERAL_FAILURE,
-    std::string("Precondition failed while aborting moveShard job ") + _jobId);
+        TRI_ERROR_SUPERVISION_GENERAL_FAILURE,
+        std::string("Precondition failed while aborting moveShard job ") + _jobId);
   }
 
   return result;
 }
+
+void MoveShard::addMoveShardToServerLock(Builder& ops) const {
+  addReadLockServer(ops, _to, _jobId);
+}
+
+void MoveShard::addMoveShardFromServerLock(Builder& ops) const {
+  if (!isSubJob()) {
+    addReadLockServer(ops, _from, _jobId);
+  }
+}
+
+void MoveShard::addMoveShardToServerUnLock(Builder& ops) const {
+  addReadUnlockServer(ops, _to, _jobId);
+}
+
+void MoveShard::addMoveShardToServerCanLock(Builder& precs) const {
+  addPreconditionServerReadLockable(precs, _to, _jobId);
+}
+
+void MoveShard::addMoveShardFromServerCanLock(Builder& precs) const {
+  if (isSubJob()) {
+    addPreconditionServerWriteLocked(precs, _from, _parentJobId);
+  } else {
+    addPreconditionServerReadLockable(precs, _from, _jobId);
+  }
+}
+
+void MoveShard::addMoveShardFromServerUnLock(Builder& ops) const {
+  if (!isSubJob()) {
+    addReadUnlockServer(ops, _from, _jobId);
+  }
+}
+
+void MoveShard::addMoveShardToServerCanUnLock(Builder& ops) const {
+  addPreconditionServerReadLocked(ops, _to, _jobId);
+}
+
+void MoveShard::addMoveShardFromServerCanUnLock(Builder& ops) const {
+  if (!isSubJob()) {
+    addPreconditionServerReadLocked(ops, _from, _jobId);
+  }
+}
+
+bool MoveShard::moveShardFinish(bool unlock, bool success, std::string const& msg) {
+  std::shared_ptr<VPackBuilder> payload;
+
+  if (unlock) {
+    payload = std::make_shared<VPackBuilder>();
+    {
+      VPackArrayBuilder env(payload.get());
+      {
+        VPackObjectBuilder trx(payload.get());
+        addReleaseShard(*payload, _shard);
+        addMoveShardToServerUnLock(*payload);
+        addMoveShardFromServerUnLock(*payload);
+      }
+      {
+        VPackObjectBuilder precs(payload.get());
+        addMoveShardFromServerCanUnLock(*payload);
+        addMoveShardToServerCanUnLock(*payload);
+      }
```

Comment:
I'm a little bit unsure here. This is probably correct, but just to clarify it for me:
I assume these preconditions should never fail. If they do fail, trying to unlock might fail the write operation midway, which we better avoid. Apart from that, unlocking unconditionally would be fine.

Did I get that right, or did I miss something?

Reply:
The preconditions check that the entity that wants to unlock actually holds the lock. I though it would be a good idea to guard against something we did not expect. I am unsure if `unlocking unconditionally` is a good idea or not.